enhancements/network/bgp-overview.md

```diff
@@ -0,0 +1,152 @@
+---
+title: bgp-overview
+authors:
+  - "@russellb"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-08-11
+last-updated: 2020-08-11
+status: informational
+see-also:
+  - https://github.com/openshift/enhancements/pull/356
+  - https://github.com/openshift/enhancements/pull/394
+---
+
+# OpenShift and BGP
+
+BGP (Border Gateway Protocol) is a dynamic routing protocol used to power a
+number of important networking use cases, including routing in the global
+internet.  For some helpful resources on learning more about BGP, see the
+[BGP Resources](#bgp-resources) section of this document.
+
+The purpose of this document is to serve as a high level overview of how BGP
+relates to OpenShift.  We will review a number of use cases, discuss how they
+could be supported, and link to existing related work.  Any future enhancements
+related to BGP should be able to rely on this one to provide some broader
+context and should continue to update this document as necessary.
+
+## Use Cases
+
+### L3 (Layer 3) Network Fabric
+
+The use case here does not require any changes to OpenShift.  It is still
+useful to call out as it is one use of BGP that is relevant to OpenShift,
+particularly with on premise clusters.
+
+It is possible to use BGP between the routers used to build the L3 network that
+OpenShift Nodes are connected to.
+
+### Dynamic Routing for Egress Traffic
+
+Some networks may have more than one router accessible from Nodes for
+availability reasons.  In this case an administrator may want to use BGP down
+to the Nodes themselves so they are aware of the multiple routing options and
+potentially use ECMP (Equal Cost Multi Path) to distribute traffic
+across each of the routers.
+
+This capability is not provided by OpenShift today.  It may be possible as a
+custom configuration, but not with all network providers.  With OpenShift-SDN,
+it should be possible to run your own BGP routing daemon on the host via a
+`DaemonSet`.
+
+OVN-Kuberenetes uses a different host network configuration where egress
+traffic is routed by Open vSwitch flows programmed by OVN.  This mode is
+referred to as "shared gateway mode".  It means that a dynamic routing daemon
+must also feed those routes into OVN configuration, so it's not as straight
+forward as just running an existing daemon on the host.
+
+A future enhancement is needed to describe how we would provide this
+functionality natively with OpenShift.
+
+### Pod Network Control Plane
+
+Many people become aware of BGP in the Kubernetes community via its use in
+[Calico][1].  The primary way BGP is used in Calico is as a control plane for
+the Pod network.  In other words, Calico uses BGP internally for Nodes to share
+the locations of Pod IP addresses in the cluster.
```

Comment:
"In this scenario, the use of BGP is a private implementation detail, and does not provide any additional user-facing features." or something like that.

Reply:
What Casey said ("does not provide any additional user-facing features") isn't accurate; it allows the pod network to be unencapsulated.

I don't know what customers are saying about BGP these days, but in the past, when a customer said "I want OpenShift to support BGP", what they actually meant (100% of the time) was "I want pod networking without VXLAN encapsulation, like Calico". They did not actually care about BGP at all; "BGP" was just a code word that meant "unencapsulated pod networking".

Reply:
In the default Calico setup, they use IP in IP tunneling - BGP really is just an internal control plane.

Reply:
Really? Did it always do that? Does it do that even if the nodes are all on the same L2 segment?

Anyway, regardless of what Calico _actually_ does, customers definitely used to tell us that they wanted us to use BGP like Calico, to avoid encapsulation.

Reply:
That sounds good, I'll use that.

Reply:
weird, i had only seen the very first comment when i sent this last reply.

Reply:
https://docs.projectcalico.org/networking/vxlan-ipip

> Calico has an option to selectively encapsulate only traffic that crosses subnet boundaries. We recommend using the cross-subnet option with IP in IP or VXLAN to minimize encapsulation overhead. Cross-subnet mode provides better performance in AWS multi-AZ deployments, Azure VNETs, and on networks where routers are used to connect pools of nodes with L2 connectivity.

Reply:
How about this ...

> In this case, the use of BGP should be considered an implementation detail of the control plane and is not a prerequisite for using BGP to satisfy other use cases.

Reply:
It's not a prerequisite for _other_ use cases, but I think _this_ use case is more interesting than the text currently implies. By using BGP, we could avoid having to encapsulate pod-to-pod traffic between nodes in the same L2 domain. Some people would like that. So, if the goal of this document is to list all of the things that we might use BGP to do, or all of the contexts in which customers might ask about BGP, then this is one of them.

Reply:
OK, I see your point.  Yes, you've captured the intent of the doc so I agree we should discuss this.  Let me think about this a bit more and propose some updated text.